WordPress/Classes/Services/SiteAssemblyService.swift

```diff
@@ -4,26 +4,178 @@ import Foundation
+            api = wpcomApi
+        } else {
+            api = WordPressComRestApi(userAgent: WPUserAgent.wordPress())
+        }
+        self.remoteService = WordPressComServiceRemote(wordPressComRestApi: api)
+
+        super.init(managedObjectContext: context)
+    }
+
+    // MARK: SiteAssemblyService
+
     private(set) var currentStatus: SiteAssemblyStatus = .idle {
         didSet {
             DispatchQueue.main.async { [weak self] in
-                guard let strongSelf = self else {
+                guard let self = self else {
                     return
                 }
-                strongSelf.statusChangeHandler?(strongSelf.currentStatus)
+                self.statusChangeHandler?(self.currentStatus)
             }
         }
     }
 
-    private(set) var statusChangeHandler: SiteAssemblyStatusChangedHandler?
+    /// This method serves as the entry point for creating an enhanced site. It consists of a few steps:
+    /// 1. The creation request is first validated.
+    /// 1. If a site passes validation, a new service invocation triggers creation.
+    /// 1. The details of the created Blog are persisted to the client.
+    ///
+    /// Previously the site's theme & tagline were set post-creation. This is now handled on the server via Headstart.
+    ///
+    /// - Parameters:
+    ///   - creationRequest:    the request with which to initiate site creation.
+    ///   - changeHandler:      this closure is invoked when site assembly status changes occur.
+    ///
+    func createSite(creationRequest: SiteCreationRequest, changeHandler: SiteAssemblyStatusChangedHandler? = nil) {
+        self.creationRequest = creationRequest
+        self.statusChangeHandler = changeHandler
 
-    // MARK: SiteAssemblyService
+        beginAssembly()
+        validatePendingRequest()
+    }
 
-    func createSite(creatorOutput assemblyInput: SiteCreatorOutput, changeHandler: SiteAssemblyStatusChangedHandler? = nil) {
-        self.statusChangeHandler = changeHandler
+    // MARK: Private behavior
+
+    private func beginAssembly() {
+        currentStatus = .inProgress
+    }
+
+    private func endFailedAssembly() {
+        currentStatus = .failed
+    }
+
+    private func endSuccessfulAssembly() {
+        // Here we designate the new site as the last used, so that it will be presented post-creation
+        if let siteUrl = createdBlog?.url {
+            RecentSitesService().touch(site: siteUrl)
+        }
+
+        currentStatus = .succeeded
+    }
+
+    private func performRemoteSiteCreation() {
+        guard let request = creationRequest else {
+            self.endFailedAssembly()
+            return
+        }
+
+        let validatedRequest = SiteCreationRequest(request: request)
+
+        remoteService.createWPComSite(request: validatedRequest) { result in
+            switch result {
+            case .success(let response):
+                guard response.success == true else {
```

Comment:
oh interesting! Why do we have to check the response.success here if the result is .success?

Reply:
Good question, and probably a sign that some additional documentation is needed.

The outer `result` is an enum returned by the [client API](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/63/files#diff-06b6b02a96de97b2573b653a36b981d0R123), with cases for both `success` & `failure`.

The former returns a value type encapsulating the response payload (which includes _another_ field named `success`); the latter returns an underlying `Error` type associated with the `failure`.

The endpoint is not documented [here](https://developer.wordpress.com/docs/api/), but it's possible that some additional detail could be found via `opengrok`.

Would it be helpful if I added some additional documentation to that effect? If so, please let me know if you'd like to see it here or in `WordPressKit`.

Reply:
Yeah maybe more documentation where the payload is defined  would help. Is it possible for the response to be success but for the "success" in the payload to be false?

Reply:
@yaelirub I have pushed an additional commit that adds some additional inline documentation clarifying the observed behavior. 

As for the payload contents / description, I think it's best if that remains in `WordPressKit`. I added thorough tests & supporting documentation via [this PR](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/63/). We can add _even_ more, but that will require another release of the framework.

Reply:
Thanks @stevebaranski !